File: src/lib/dates.ts
import { khmerNum } from "./chhankitek";
import { monthsOf, t, weekdaysFull, type Lang } from "./i18n";

export function isoOf(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

export function fromIso(iso: string): Date {
  const [y, m, d] = iso.split("-").map(Number);
  return new Date(y, (m ?? 1) - 1, d ?? 1);
}

export function startOfMonth(d: Date): Date {
  return new Date(d.getFullYear(), d.getMonth(), 1);
}

export function addDays(d: Date, n: number): Date {
  const x = new Date(d.getFullYear(), d.getMonth(), d.getDate());
  x.setDate(x.getDate() + n);
  return x;
}

export function addMonths(d: Date, n: number): Date {
  const x = new Date(d.getFullYear(), d.getMonth() + n, 1);
  const day = Math.min(d.getDate(), daysInMonth(x));
  x.setDate(day);
  return x;
}

export function daysInMonth(d: Date): number {
  return new Date(d.getFullYear(), d.getMonth() + 1, 0).getDate();
}

export function sameDay(a: Date, b: Date): boolean {
  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate();
}

export function sameMonth(a: Date, b: Date): boolean {
  return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth();
}

export function monthGrid(month: Date, weekStartsOn: 0 | 1 | 6): Date[] {
  const first = startOfMonth(month);
  const firstDow = first.getDay();
  const offset = (firstDow - weekStartsOn + 7) % 7;
  const start = addDays(first, -offset);
  return Array.from({ length: 42 }, (_, i) => addDays(start, i));
}

export function todayIso(): string {
  return isoOf(new Date());
}

export function roundTime5(): string {
  const e = new Date();
  const t = e.getMinutes();
  const n = Math.ceil(t / 5) * 5;
  const roll = n >= 60;
  const h = (e.getHours() + Number(roll)) % 24;
  const m = roll ? 0 : n;
  return `${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}`;
}

export function formatTime(hhmm: string): string {
  if (!hhmm) return "";
  const [h, m] = hhmm.split(":").map(Number);
  return `${String(h ?? 0).padStart(2, "0")}:${String(m ?? 0).padStart(2, "0")}`;
}

export function formatTime12(hhmm: string): string {
  if (!hhmm) return "";
  const [h, m] = hhmm.split(":").map(Number);
  if (!Number.isFinite(h) || !Number.isFinite(m)) return hhmm;
  const am = h < 12;
  return `${h % 12 || 12}:${String(m).padStart(2, "0")} ${am ? "AM" : "PM"}`;
}

export function formatMonthTitle(d: Date, lang: Lang) {
  const name = monthsOf(lang)[d.getMonth()];
  const y = d.getFullYear();
  return lang === "en" ? `${name} ${y}` : `${name} ${khmerNum(y)}`;
}

export function nowHm() {
  const e = new Date();
  return `${String(e.getHours()).padStart(2, "0")}:${String(e.getMinutes()).padStart(2, "0")}`;
}

export function reminderChipLabel(date: string, time: string, lang: Lang): string {
  if (!date) return t(lang, "addReminder");
  const today = todayIso();
  const d = fromIso(date);
  const months = monthsOf(lang);
  const wdays = weekdaysFull(lang);
  const dayMonth = `${d.getDate()} ${months[d.getMonth()]}`;
  const wday = wdays[d.getDay()];
  const tm = time ? formatTime12(time) : "";
  if (date === today) return tm ? `${t(lang, "today")}, ${tm}` : t(lang, "today");
  if (date.slice(0, 7) === today.slice(0, 7)) return tm ? `${dayMonth}, ${tm}` : dayMonth;
  const diff = Math.abs(d.getTime() - fromIso(today).getTime());
  if (diff < 7 * 86400000) return tm ? `${wday}, ${tm}` : wday;
  return tm ? `${dayMonth}, ${tm}` : dayMonth;
}

export function eventWhenLabel(date: string, startTime: string | undefined, allDay: boolean | undefined, lang: Lang): string {
  if (!date) return "";
  const d = fromIso(date);
  const today = fromIso(todayIso());
  const diff = Math.round((d.getTime() - today.getTime()) / 86400000);
  const months = monthsOf(lang);
  const wdays = weekdaysFull(lang);
  const label =
    diff === 0 ? t(lang, "today") : diff > 0 && diff < 7 ? wdays[d.getDay()] : `${d.getDate()} ${months[d.getMonth()]}`;
  if (allDay || !startTime) return label;
  return `${label}, ${formatTime12(startTime)}`;
}

export function eventWhenTone(date: string, done?: boolean): "" | "is-overdue" | "is-today" {
  if (done || !date) return "";
  const today = todayIso();
  if (date < today) return "is-overdue";
  if (date === today) return "is-today";
  return "";
}

export function showNativePicker(el: HTMLInputElement | null) {
  if (!el) return;
  try {
    if (typeof el.showPicker === "function") {
      el.showPicker();
      return;
    }
  } catch {
    /* fall through */
  }
  el.focus();
  el.click();
}

export function joinTitleNotes(title: string, notes?: string) {
  return title && notes ? `${title}\n${notes}` : notes || title;
}

export function splitTitleNotes(raw: string) {
  const text = raw.replace(/\r\n/g, "\n").trim();
  const n = text.indexOf("\n");
  if (n === -1) return { title: text, notes: "" };
  return { title: text.slice(0, n).trim(), notes: text.slice(n + 1) };
}

File: khmer_calendar/lib/dates.dart
import 'calendar/chhankitek.dart';
import 'i18n.dart';

const calendarStartYear = 1900;
const calendarEndYear = 2100;

int monthIndexOf(DateTime d) => (d.year - calendarStartYear) * 12 + (d.month - 1);

int monthCount() => (calendarEndYear - calendarStartYear + 1) * 12;

DateTime monthFromIndex(int i) => DateTime(calendarStartYear, 1 + i, 1);

int dayIndexOf(DateTime d) => DateTime.utc(d.year, d.month, d.day).difference(DateTime.utc(calendarStartYear, 1, 1)).inDays;

int dayCount() => DateTime.utc(calendarEndYear, 12, 31).difference(DateTime.utc(calendarStartYear, 1, 1)).inDays + 1;

DateTime dayFromIndex(int i) => DateTime(calendarStartYear, 1, 1 + i);

String isoOf(DateTime d) {
  final y = d.year.toString().padLeft(4, '0');
  final m = d.month.toString().padLeft(2, '0');
  final day = d.day.toString().padLeft(2, '0');
  return '$y-$m-$day';
}

DateTime fromIso(String iso) {
  final p = iso.split('-').map(int.parse).toList();
  return DateTime(p[0], p.length > 1 ? p[1] : 1, p.length > 2 ? p[2] : 1);
}

DateTime startOfMonth(DateTime d) => DateTime(d.year, d.month, 1);

DateTime addDays(DateTime d, int n) => DateTime(d.year, d.month, d.day + n);

int daysInMonth(DateTime d) => DateTime(d.year, d.month + 1, 0).day;

DateTime addMonths(DateTime d, int n) {
  final x = DateTime(d.year, d.month + n, 1);
  final day = d.day.clamp(1, daysInMonth(x));
  return DateTime(x.year, x.month, day);
}

bool sameDay(DateTime a, DateTime b) => a.year == b.year && a.month == b.month && a.day == b.day;
bool sameMonth(DateTime a, DateTime b) => a.year == b.year && a.month == b.month;

List<DateTime> monthGrid(DateTime month, int weekStartsOn) {
  final first = startOfMonth(month);
  final firstDow = first.weekday % 7;
  final offset = (firstDow - weekStartsOn + 7) % 7;
  final start = addDays(first, -offset);
  return List.generate(42, (i) => addDays(start, i));
}

String todayIso() => isoOf(DateTime.now());

String formatTime12(String hhmm) {
  if (hhmm.isEmpty) return '';
  final p = hhmm.split(':').map(int.tryParse).toList();
  final h = p[0] ?? 0;
  final m = p.length > 1 ? (p[1] ?? 0) : 0;
  final am = h < 12;
  return '${h % 12 == 0 ? 12 : h % 12}:${m.toString().padLeft(2, '0')} ${am ? 'AM' : 'PM'}';
}

String formatMonthTitle(DateTime d, Lang lang) {
  final name = monthsOf(lang)[d.month - 1];
  final y = d.year;
  return lang == Lang.en ? '$name $y' : '$name ${khmerNum(y)}';
}

String eventWhenLabel(String date, String? startTime, bool? allDay, Lang lang) {
  if (date.isEmpty) return '';
  final d = fromIso(date);
  final today = fromIso(todayIso());
  final diff = d.difference(today).inDays;
  final months = monthsOf(lang);
  final wdays = weekdaysFull(lang);
  final label = diff == 0
      ? t(lang, 'today')
      : (diff > 0 && diff < 7)
          ? wdays[d.weekday % 7]
          : '${d.day} ${months[d.month - 1]}';
  if (allDay == true || startTime == null || startTime.isEmpty) return label;
  return '$label, ${formatTime12(startTime)}';
}

String newId() {
  final now = DateTime.now().millisecondsSinceEpoch.toRadixString(36);
  final r = (DateTime.now().microsecondsSinceEpoch % 100000).toRadixString(36);
  return 'evt-$now-$r';
}
